### PR TITLE
[RFC] add display=msgsep flag to separate messages and avoid full-screen scrolling

### PR DESCRIPTION
**Pull request by bfredl**
Instead of scrolling the entire screen like a dumb terminal, only scroll the messages that has been shown so far. It looks best with `laststatus=2`, to give a visual marker of the scrolled region. (of course a new solid line could be added for this purpose). Pager is unaffected, and scrolls the entire screen as usual.

Not sure if/how we want to integrate this though:

- Make it the default, and deal with the pitchforks.
- Add an option. But options are evil, add complexity and all that.
- Do nothing. The vintage full terminal scroll obviously adds to the experience of a real UNIX editor.

### CODE REVIEW

**Comment by justinmk** [comment]
> It looks best with laststatus=2, to give a visual marker of the scrolled region. (of course a new solid line could be added for this purpose)

I would suggest always showing some sort of line (the statusline isn't really relevant in this state), one of these:

```
─       hh      2500    9472    BOX DRAWINGS LIGHT HORIZONTAL
━       HH      2501    9473    BOX DRAWINGS HEAVY HORIZONTAL
```

> Not sure if/how we want to integrate this though:

Definitely it should be the default. 

Probably it should be an option, if only to avoid churn in tests. Could be a new flag in `'display'`.

**Comment by bfredl** [comment]
I thought of `shortmess` but `display` is probably better, as we can make keyword and not single char (which has higher risk of vim-dev conflict)

I will add a dedicated line. If we use unicode char it needs to be overridable, what if the users favourite font looks ugly with this char. Probably should add `fillchars` for this.

**Comment by bfredl** [comment]
Added `display=msgsep` flag, `filchars=msgsep: ` (blank) default, and `hi link MsgSeparator StatusLine` highlight.

**Comment by bfredl** [comment]
Interesting ASAN failure. We probably need to add some check to not enter the debugger after start of teardown.

```
[ RUN      ] Ex commands coloring support does not prevent mapping error from cancelling prompt: ERR
==================== File /home/travis/build/neovim/neovim/build/log/ubsan.21230 ====================
= =================================================================
= ==21230==ERROR: AddressSanitizer: heap-use-after-free on address 0x604000000398 at pc 0x000000a24723 bp 0x7fff7c971fc0 sp 0x7fff7c971fb8
= READ of size 8 at 0x604000000398 thread T0
=     #0 0xa24722 in QUEUE_EMPTY /home/travis/build/neovim/neovim/src/nvim/lib/queue.h:46:18
=     #1 0xa24c9d in multiqueue_empty /home/travis/build/neovim/neovim/src/nvim/event/multiqueue.c:167:10
=     #2 0xa26b30 in multiqueue_process_events /home/travis/build/neovim/neovim/src/nvim/event/multiqueue.c:147:11
=     #3 0xa2079f in loop_poll_events /home/travis/build/neovim/neovim/src/nvim/event/loop.c:66:3
=     #4 0x123bff0 in os_breakcheck /home/travis/build/neovim/neovim/src/nvim/os/input.c:157:5
=     #5 0xd4b9bb in vgetorpeek /home/travis/build/neovim/neovim/src/nvim/getchar.c:1687:11
=     #6 0xd4857b in vpeekc /home/travis/build/neovim/neovim/src/nvim/getchar.c:1537:10
=     #7 0xd5acec in char_avail /home/travis/build/neovim/neovim/src/nvim/getchar.c:1564:12
=     #8 0x1714a22 in state_enter /home/travis/build/neovim/neovim/src/nvim/state.c:40:9
=     #9 0xbdbf01 in command_line_enter /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:385:3
=     #10 0xbd5520 in getcmdline /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:1980:10
=     #11 0xbdf581 in getcmdline_prompt /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:2018:29
=     #12 0xabb88d in do_debug /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:192:25
=     #13 0xabeb27 in dbg_check_breakpoint /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:475:7
=     #14 0xb1304c in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:1498:3
=     #15 0xb05da3 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:609:20
=     #16 0xb0c0b5 in do_cmdline_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:280:10
=     #17 0xf58eda in free_all_mem /home/travis/build/neovim/neovim/src/nvim/memory.c:599:3
=     #18 0x125288c in mch_exit /home/travis/build/neovim/neovim/src/nvim/os_unix.c:153:3
=     #19 0x1029093 in exit_event /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/channel.c:539:5
=     #20 0xa26bba in multiqueue_process_events /home/travis/build/neovim/neovim/src/nvim/event/multiqueue.c:150:7
=     #21 0xa2079f in loop_poll_events /home/travis/build/neovim/neovim/src/nvim/event/loop.c:66:3
=     #22 0x123fd56 in input_poll /home/travis/build/neovim/neovim/src/nvim/os/input.c:349:3
=     #23 0x123b4cf in inbuf_poll /home/travis/build/neovim/neovim/src/nvim/os/input.c:372:24
=     #24 0x123b335 in os_inchar /home/travis/build/neovim/neovim/src/nvim/os/input.c:115:18
=     #25 0x1714b5e in state_enter /home/travis/build/neovim/neovim/src/nvim/state.c:55:13
=     #26 0xbdbf01 in command_line_enter /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:385:3
=     #27 0xbd5520 in getcmdline /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:1980:10
=     #28 0xbdf581 in getcmdline_prompt /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:2018:29
=     #29 0xabb88d in do_debug /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:192:25
=     #30 0xabeb27 in dbg_check_breakpoint /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:475:7
=     #31 0xb1304c in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:1498:3
=     #32 0xb05da3 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:609:20
=     #33 0xb0c0b5 in do_cmdline_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:280:10
=     #34 0xabe1f6 in ex_debug /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:427:3
=     #35 0xb23b3c in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2242:5
=     #36 0xb05da3 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:609:20
=     #37 0x10afa37 in nv_colon /home/travis/build/neovim/neovim/src/nvim/normal.c:4558:18
=     #38 0x109a01a in normal_execute /home/travis/build/neovim/neovim/src/nvim/normal.c:1136:3
=     #39 0x1714d89 in state_enter /home/travis/build/neovim/neovim/src/nvim/state.c:67:26
=     #40 0x1056f64 in normal_enter /home/travis/build/neovim/neovim/src/nvim/normal.c:466:3
=     #41 0xe4983c in main /home/travis/build/neovim/neovim/src/nvim/main.c:572:3
=     #42 0x2b54ce7aaf44 in __libc_start_main /build/eglibc-ripdx6/eglibc-2.19/csu/libc-start.c:287
=     #43 0x44d39b in _start (/home/travis/build/neovim/neovim/build/bin/nvim+0x44d39b)
= 
= 0x604000000398 is located 8 bytes inside of 48-byte region [0x604000000390,0x6040000003c0)
= freed by thread T0 here:
=     #0 0x4ee232 in __interceptor_free /local/mnt/workspace/tmp/ubuntu_rel/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:47:3
=     #1 0xf56654 in xfree /home/travis/build/neovim/neovim/src/nvim/memory.c:133:3
=     #2 0xa24645 in multiqueue_free /home/travis/build/neovim/neovim/src/nvim/event/multiqueue.c:126:3
=     #3 0xa21eae in loop_close /home/travis/build/neovim/neovim/src/nvim/event/loop.c:138:3
=     #4 0xe46714 in event_teardown /home/travis/build/neovim/neovim/src/nvim/main.c:178:10
=     #5 0x125284c in mch_exit /home/travis/build/neovim/neovim/src/nvim/os_unix.c:145:8
=     #6 0x1029093 in exit_event /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/channel.c:539:5
=     #7 0xa26bba in multiqueue_process_events /home/travis/build/neovim/neovim/src/nvim/event/multiqueue.c:150:7
=     #8 0xa2079f in loop_poll_events /home/travis/build/neovim/neovim/src/nvim/event/loop.c:66:3
=     #9 0x123fd56 in input_poll /home/travis/build/neovim/neovim/src/nvim/os/input.c:349:3
=     #10 0x123b4cf in inbuf_poll /home/travis/build/neovim/neovim/src/nvim/os/input.c:372:24
=     #11 0x123b335 in os_inchar /home/travis/build/neovim/neovim/src/nvim/os/input.c:115:18
=     #12 0x1714b5e in state_enter /home/travis/build/neovim/neovim/src/nvim/state.c:55:13
=     #13 0xbdbf01 in command_line_enter /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:385:3
=     #14 0xbd5520 in getcmdline /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:1980:10
=     #15 0xbdf581 in getcmdline_prompt /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:2018:29
=     #16 0xabb88d in do_debug /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:192:25
=     #17 0xabeb27 in dbg_check_breakpoint /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:475:7
=     #18 0xb1304c in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:1498:3
=     #19 0xb05da3 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:609:20
=     #20 0xb0c0b5 in do_cmdline_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:280:10
=     #21 0xabe1f6 in ex_debug /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:427:3
=     #22 0xb23b3c in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2242:5
=     #23 0xb05da3 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:609:20
=     #24 0x10afa37 in nv_colon /home/travis/build/neovim/neovim/src/nvim/normal.c:4558:18
=     #25 0x109a01a in normal_execute /home/travis/build/neovim/neovim/src/nvim/normal.c:1136:3
=     #26 0x1714d89 in state_enter /home/travis/build/neovim/neovim/src/nvim/state.c:67:26
=     #27 0x1056f64 in normal_enter /home/travis/build/neovim/neovim/src/nvim/normal.c:466:3
=     #28 0xe4983c in main /home/travis/build/neovim/neovim/src/nvim/main.c:572:3
=     #29 0x2b54ce7aaf44 in __libc_start_main /build/eglibc-ripdx6/eglibc-2.19/csu/libc-start.c:287
= 
= previously allocated by thread T0 here:
=     #0 0x4ee573 in malloc /local/mnt/workspace/tmp/ubuntu_rel/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:67:3
=     #1 0xf562e1 in try_malloc /home/travis/build/neovim/neovim/src/nvim/memory.c:87:15
=     #2 0xf56509 in xmalloc /home/travis/build/neovim/neovim/src/nvim/memory.c:121:15
=     #3 0xa23bf2 in multiqueue_new /home/travis/build/neovim/neovim/src/nvim/event/multiqueue.c:103:20
=     #4 0xa240e6 in multiqueue_new_child /home/travis/build/neovim/neovim/src/nvim/event/multiqueue.c:97:10
=     #5 0xa1edc8 in loop_init /home/travis/build/neovim/neovim/src/nvim/event/loop.c:26:23
=     #6 0xe4655a in event_init /home/travis/build/neovim/neovim/src/nvim/main.c:146:3
=     #7 0xe46ede in main /home/travis/build/neovim/neovim/src/nvim/main.c:261:3
=     #8 0x2b54ce7aaf44 in __libc_start_main /build/eglibc-ripdx6/eglibc-2.19/csu/libc-start.c:287
= 
= SUMMARY: AddressSanitizer: heap-use-after-free /home/travis/build/neovim/neovim/src/nvim/lib/queue.h:46:18 in QUEUE_EMPTY
= Shadow bytes around the buggy address:
=   0x0c087fff8020: fa fa 00 00 00 00 00 00 fa fa 00 00 00 00 07 fa
=   0x0c087fff8030: fa fa fd fd fd fd fd fd fa fa fd fd fd fd fd fd
=   0x0c087fff8040: fa fa fd fd fd fd fd fd fa fa fd fd fd fd fd fd
=   0x0c087fff8050: fa fa 00 00 00 00 00 04 fa fa 00 00 00 00 00 04
=   0x0c087fff8060: fa fa fd fd fd fd fd fd fa fa fd fd fd fd fd fd
= =>0x0c087fff8070: fa fa fd[fd]fd fd fd fd fa fa fd fd fd fd fd fd
=   0x0c087fff8080: fa fa 00 00 00 00 00 fa fa fa 00 00 00 00 00 fa
=   0x0c087fff8090: fa fa fd fd fd fd fd fd fa fa 00 00 00 00 00 00
=   0x0c087fff80a0: fa fa 00 00 00 00 00 fa fa fa 00 00 00 00 00 fa
=   0x0c087fff80b0: fa fa fd fd fd fd fd fd fa fa fd fd fd fd fd fa
=   0x0c087fff80c0: fa fa fd fd fd fd fd fd fa fa fd fd fd fd fd fa
= Shadow byte legend (one shadow byte represents 8 application bytes):
=   Addressable:           00
=   Partially addressable: 01 02 03 04 05 06 07 
=   Heap left redzone:       fa
=   Freed heap region:       fd
=   Stack left redzone:      f1
=   Stack mid redzone:       f2
=   Stack right redzone:     f3
=   Stack after return:      f5
=   Stack use after scope:   f8
=   Global redzone:          f9
=   Global init order:       f6
=   Poisoned by user:        f7
=   Container overflow:      fc
=   Array cookie:            ac
=   Intra object redzone:    bb
=   ASan internal:           fe
=   Left alloca redzone:     ca
=   Right alloca redzone:    cb
= ==21230==ABORTING
=====================================================================================================
test/helpers.lua:95: assertion failed!
```

**Comment by justinmk** [comment]
> fillchars=msgsep: (blank) default

It should be visible in some form by default (either color or fillchars).

**Comment by bfredl** [comment]
It is, `hi link MsgSeparator StatusLine`, gives a nice solid line in same color as status line.

File: src/nvim/options.lua [hunk 1]
```diff
@@ -642,7 +642,7 @@ return {
       vim=true,
       redraw={'all_windows'},
       varname='p_dy',
-      defaults={if_true={vi="", vim="lastline"}}
+      defaults={if_true={vi="", vim="lastline,msgsep"}}
```
Comment (justinmk) [inline src/nvim/options.lua]: also mention in `vim_diff.txt`

**Comment by bfredl** [comment]
> also mention in vim_diff.txt

Yes, I haven't started with docs yet :) Is the name `msgsep` OK?

**Comment by bfredl** [comment]
Added tests and docs, marking RFC.

**Comment by justinmk** [comment]
what's going on with asan?  Is there any way to avoid this? Don't want to increase false failures in CI...

https://api.travis-ci.org/v3/job/358522097/log.txt

```
[32m[2m[ RUN      ][0m[0m :Man man.lua: highlight_line() clears backspaces from text and adds highlights: [2m5.59 ms[0m [1m[32mOK[0m[0m
==================== File /home/travis/build/neovim/neovim/build/log/ubsan.1567 ====================
= =================================================================
= ==1567==ERROR: AddressSanitizer: SEGV on unknown address 0x000001b458f0 (pc 0x00000044e641 bp 0x000000000001 sp 0x7ffcd7c90880 T0)
= ==1567==The signal is caused by a WRITE memory access.
=     #0 0x44e640 in atomic_compare_exchange_strong<__sanitizer::atomic_uint8_t> /local/mnt/workspace/tmp/ubuntu_rel/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_atomic_clang.h:90:10
=     #1 0x44e640 in AtomicallySetQuarantineFlagIfAllocated /local/mnt/workspace/tmp/ubuntu_rel/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/asan_allocator.cc:520
=     #2 0x44e640 in __asan::Allocator::Deallocate(void*, unsigned long, __sanitizer::BufferedStackTrace*, __asan::AllocType) /local/mnt/workspace/tmp/ubuntu_rel/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/asan_allocator.cc:597
=     #3 0x4ee5b6 in __interceptor_free /local/mnt/workspace/tmp/ubuntu_rel/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:50:3
=     #4 0xf5abf4 in xfree /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:133:3
=     #5 0x11a94a8 in free_string_option /home/travis/build/neovim/neovim/build/../src/nvim/option.c:2233:5
=     #6 0x11a8992 in free_all_options /home/travis/build/neovim/neovim/build/../src/nvim/option.c:947:9
=     #7 0xf5d551 in free_all_mem /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:617:3
=     #8 0x1257ecc in mch_exit /home/travis/build/neovim/neovim/build/../src/nvim/os_unix.c:153:3
...
```

**Comment by justinmk** [comment]
The ASAN failure in my previous comment isn't related to this PR, it happened on #8184 also. https://api.travis-ci.org/v3/job/358244576/log.txt

**Comment by bfredl** [comment]
Seems at any point `os_breakcheck` is called, nvim could decide to exit (in the clean fashion, which IIRC will include arbitrary event processing), do we really want that?

As the ASAN failures look unrelated, should we merge this? If this causes trouble/backlash for master-followers, there is ample time to change the default back before the release.

File: src/nvim/syntax.c [hunk 2]
```diff
@@ -5999,6 +5999,7 @@ static const char *highlight_init_both[] = {
   "default link QuickFixLine Search",
   "default link Substitute Search",
   "default link Whitespace NonText",
+  "default link MsgSeparator StatusLine",
```
Comment (justinmk) [inline src/nvim/syntax.c]: the others are in alphabetical order

**Review by justinmk — APPROVED** [review]
+1 to merge

**Comment by blueyed** [comment]
@bfredl 
Thanks for this very nice feature!

**Pull request merged**